Allow to strip root folder. Closes #588, closes #5433

glassez opened the pull request:
As @sledgehammer999 has no time to maintain #4081 up to date, I offer to your attention this PR.

Bhaalspawn commented:
What is this?

glassez commented:
@Bhaalspawn, you can see more info in referenced issues: #588, #5433 and #4081.

LordNyriox commented:
@glassez:  Speaking of maintenance, this PR needs a rebase.  

Thanks.

glassez commented:
PR rebased.
@sledgehammer999, Anyone still interested in this?
Please test it.

Matth7878 commented:
Nobody with the skill is willing to test/review this pull request ? I'm asking out of curiosity because it seems to be that it is a feature a lot of people are waiting for yet it feels like it's not getting any attention. (That's not a critic ! :-))

sledgehammer999 commented:
@glassez please keep it open. I'll review it eventually. This week will be hard though. Also ignore @LordNyriox's requests for rebase.

@LordNyriox please, stop telling people to rebase their PRs. I know that you mean well. But in case I won't review their PR for some time they will constant nags from you for rebasing it. I suspect that this will annoy them to constantly have to update their code, even if they don't say something. If I want to review some PR and it needs rebasing, I'll tell them so myself. Thank you.

glassez commented:
@sledgehammer999 maybe you look in there now, apparently, you have the time?
Although, I'm not sure I can answer your questions about it, because I already forgot what I was doing here...

sledgehammer999 commented:
I am keeping this open in a tab. However, #5214 takes the most precedence.
If you're able to rebase this in the meantime, that would be great.

glassez commented:
>If you're able to rebase this in the meantime, that would be great.

Rebased.

sledgehammer999 commented on src/gui/addnewtorrentdialog.ui:
```diff
@@ -192,6 +192,26 @@
        </spacer>
       </item>
+      <item row="0" column="0">
+       <widget class="QCheckBox" name="start_torrent_cb">
```
I don't see this utilized in this commit. Also the naming scheme is wrong.

sledgehammer999 commented on src/base/bittorrent/torrentinfo.cpp:
```diff
@@ -276,22 +276,10 @@ TorrentInfo::PieceRange TorrentInfo::filePieces(int fileIndex) const
 {
     if (!isValid()) return;
-    nativeInfo()->rename_file(index, newPath.toStdString());
+    nativeInfo()->rename_file(index, Utils::Fs::toNativePath(newPath).toStdString());
```
I think libtorrent generally expects paths using `/` as dir separator...

glassez replied:
Apparently, I checked that, or it was in the old code. In any case, it did not cause problems, at least.

sledgehammer999 commented on src/base/bittorrent/torrentinfo.cpp:
```diff
@@ -305,10 +293,22 @@ int BitTorrent::TorrentInfo::fileIndex(const QString& fileName) const
-    m_nativeInfo->remap_files(fileStorage);
+    if (filesCount() <= 1) return;
+
+    libtorrent::file_storage files = m_nativeInfo->files();
```
Are you sure you didn't mean orig_files() here?
In any case, what is the problem that the lines below try to solve?

sledgehammer999 commented on src/base/bittorrent/torrentinfo.cpp:
```diff
@@ -305,10 +293,22 @@ int BitTorrent::TorrentInfo::fileIndex(const QString& fileName) const
+    if (files.name() != testName) {
+        files.set_name(testName);
+        for (int i = 0; i < files.num_files(); ++i)
+            files.rename_file(i, files.file_path(i));
```
Doesn't "set_name" take care of the renaming too?

glassez replied:
As far as I remember when you rename root folder in AddNewTorrentDialog the files become independent from storage (since it renamed via rename_file and storage still has its original name). We bind it together here.

glassez replied:
Oops... I Think I remember something... And if that's not obsession, then here is the solution to more General problems than this PR has.
@qbittorrent, can someone do the following test and confirm the problem?
1. Start adding some multifile torrent via AddTorrentDialog
2. Rename its root folder (in torrent content widget)
3. Let torrent to download some files
4. Do "Set Location"

If the problem I think about really exists then qBittorrent will show you the new torrent save path but its files stay in old place.

thalieht replied:
@glassez works fine for me.

glassez replied:
@sledgehammer999, OK, I remembered this. In this case, the files are detached only from the storage name, so they are affected by move_storage, but not affected by the storage renaming. That is why this code is here. Otherwise we will not be able to delete the root folder by renaming the storage to an empty string.
@thalieht, thank you for testing.

LordNyriox commented:
@sledgehammer999, @evsh, @Chocobo1, @glassez, @magao:  
On the topic of my prior rebase-requesting habit:  I do not plan to post any more rebase requests in the foreseeable future.  In fact, I deliberately stopped posting them almost a month ago.  

Of course, I will still be lurking around the pull requests because I find the way coding is handled for the qBittorrent project fascinating.  However, I will not post as often—since I generally seem to ruin any work I participate in, in one way or another.  

Thank you again for all your patience.  
– Lord Nyriox

glassez commented:
Damn, this PR is for many years. It is based on outdated code, so it contains the wrong names, obsolete approaches, etc. Easier to accept it as is, and then correct its shortcomings in a separate commit. If we don't do it now, it will die without being born.

glassez commented:
I mean I have no time for this PR, at least until I finish #6627. Although, I had already planned other important work after.

glassez commented:
@sledgehammer999 I watched it again and I realized that I have no clue what it is like! After many time I'm not in the subject. Perhaps (most likely) something got mangled during rebasing... because it was made when the affected code is substantially different from the present. 
In short, I don't want to get into it and maintain it. This abuse of my mind! If it was in my best interest and I used this feature for my purposes, it would be always in the correct state... But now I can't (and don't want) to deal with this more. As we say: "the Train has left!".
I'm sorry!

glassez commented:
@sledgehammer999, Okay, I'm a little cold... I've addressed your review. Let's merge it.
I hope this was the last headache from the legacy development process approach in this project.

WolfganP commented:
Just wanted to thank you guys for this PR. This 'avoid to create root folder' functionality was a big pain point for my usage style since converting from uTorrent. Thanks again.

sledgehammer999 commented:
Thx for the followup @glassez.
Thruth is I didn't get much into this feature. I am going to merge it anyway. It doesn't make sense to delay it further in order to make it "perfect".
@ others : Let's consider this feature experimental for the time being. It might be broken in a subtle way. Hopefully by merging this, beta testers will start reporting problems.

bobbintb commented:
Finally! THANK YOU!

thalieht commented:
This closes a few issues: #588 #6415 #5636 #5699 #4898 for your convenience.

bobbintb commented:
One question: will this be in the webUI? I use qbittorrent on a headless server so hopefully it is or will be soon so I can finally be down to just one torrent client.

glassez commented:
@ngosang, @Chocobo1, @buinsky?

WolfganP commented:
> @bobbintb One question: will this be in the webUI? I use qbittorrent on a headless server so hopefully it is or will be soon so I can finally be down to just one torrent client.

I have exactly the same question as I'm a qb-nox user. Will this be available as an option on the webui Add Torrent Link popup?

glassez commented:
>I have exactly the same question as I'm a qb-nox user. Will this be available as an option on the webui Add Torrent Link popup?

@Piccirello, @Chocobo1, @ngosang?

KaXaSA commented:
[Question] Will this also be implemented in the **Automated RSS Downloader**?
